canopeum_backend/canopeum_backend/views.py

```diff
@@ -974,71 +979,103 @@ def patch(self, request: Request, batchId):
-            for specie in parsed_species:
-                batch.add_specie_by_id(specie["id"], specie.get("quantity", 0))
-            for supported_specie_id in parsed_supported_species_ids:
-                batch.add_supported_specie_by_id(supported_specie_id)
+            self._update_batch_mappings(batch, parsed_data, asset_instances)
 
         if errors:
             return Response(data={"errors": errors}, status=status.HTTP_400_BAD_REQUEST)
 
         return Response(batch_serializer.data)
 
+    def _parse_request_data(self, request):
+        try:
+            return {
+                "fertilizer_ids": request.data.getlist("fertilizer_ids", []),
+                "mulch_layer_ids": request.data.getlist("mulch_layer_ids", []),
+                "seeds": [json.loads(seed) for seed in request.data.getlist("seeds", [])],
+                "species": [json.loads(specie) for specie in request.data.getlist("species", [])],
+                "supported_specie_ids": request.data.getlist("supported_specie_ids", []),
+            }, None
+        except json.JSONDecodeError as e:
+            return None, str(e)
+
+    def _handle_assets(self, uploaded_files, batch):
+        asset_errors = []
+        asset_instances = []
+
+        existing_assets = list(BatchAsset.objects.filter(batch=batch).select_related("asset"))
+        new_file_names = {file.name for file in uploaded_files}
+
+        for ba in existing_assets:
+            if ba.asset.asset.name not in new_file_names:
+                ba.delete()
+
+        for file in uploaded_files:
+            matching_existing = next(
+                (ba.asset for ba in existing_assets if ba.asset.asset.name == file.name), None
+            )
+            if matching_existing:
+                asset_instances.append(matching_existing)
+                continue
+
+            asset_serializer = AssetSerializer(data={"asset": file})
+            if not asset_serializer.is_valid():
+                asset_errors.append(asset_serializer.errors)
+            else:
+                asset_instances.append(asset_serializer.save())
+
+        return asset_instances, asset_errors
+
+    def _handle_sponsor(self, request, batch):
+        sponsor_data = {
+            "name": request.data.get("sponsor_name"),
+            "url": request.data.get("sponsor_website_url"),
+        }
+        if request.data.get("sponsor_logo") is not None:
+            sponsor_data["logo"] = {"asset": request.data.get("sponsor_logo")}
+
+        sponsor_serializer = BatchSponsorSerializer(batch.sponsor, data=sponsor_data, partial=True)
+        if not sponsor_serializer.is_valid():
+            return None, sponsor_serializer.errors
+        return sponsor_serializer.save(), []
+
+    def _update_batch_mappings(self, batch, parsed_data, asset_instances):
+        for asset in asset_instances:
+            BatchAsset.objects.create(batch=batch, asset=asset)
+
+        Batchfertilizer.objects.filter(batch=batch).delete()
```

Comment:
Let's keep this comment
```suggestion
        # Less efficient, but so much easier to just remove all then recreate mappings.
        Batchfertilizer.objects.filter(batch=batch).delete()
```